pennylane/plugins/default_qubit.py

```diff
@@ -240,10 +240,38 @@ def reset(self):
         self._state[0] = 1
         self._pre_rotated_state = self._state
 
-    def probability(self, wires=None):
+    def analytic_probability(self, wires=None):
+        """Return the (marginal) analytic probability of each computational basis."""
         if self._state is None:
             return None
 
         wires = wires or range(self.num_wires)
+
         prob = self.marginal_prob(np.abs(self._state) ** 2, wires)
         return prob
+
+    def _estimate_probability(self, wires=None):
+        # non-analytic mode, estimate the probability from the generated samples
+
+        # consider only the requested wires
+        wires = np.hstack(wires)
+
+        samples = self._samples[:, np.array(wires)]
+
+        # convert samples from a list of 0, 1 integers, to base 10 representation
+        unraveled_indices = [2] * len(wires)
+        indices = np.ravel_multi_index(samples.T, unraveled_indices)
+
+        # count the basis state occurrences, and construct the probability vector
+        basis_states, counts = np.unique(indices, return_counts=True)
+        prob = np.zeros([len(wires) ** 2], dtype=np.float64)
+        prob[basis_states] = counts / self.shots
+        return prob
+
+    def probability(self, wires=None):
+        wires = wires or range(self.num_wires)
+
+        if hasattr(self, "analytic") and self.analytic:
```

Comment:
`default.qubit` being a state vector simulator has the `analytic` attribute defined. As per our discussion, I'd be in for having `hasattr(self, "analytic")` as part of the condition once this is moved to `QubitDevice` (or leaving the try-except block seems a comparably good solution as well to me).
```suggestion
        if self.analytic:
```

Reply:
Agree, this should be kept when moved to `QubitDevice`.

Reply:
Sure, I just added the `hasattr` here so that it wouldn't be forgotten when transferring it to `QubitDevice`, but it's a bit useless, and perhaps unnecessary, in `DefaultQubit`.